### PR TITLE
Reconciliation extension

### PR DESCRIPTION
**Pull request by chbatey**
Basic reconciliation that we can add to.

References #578

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -44,6 +44,13 @@ lazy val core = (project in file("core"))
     testOptions in Test ++= Seq(Tests.Argument(TestFrameworks.ScalaTest, "-o")))
   .configs(MultiJvm)
 
+lazy val reconciler = (project in file("reconciler"))
+  .enablePlugins(Common, AutomateHeaderPlugin, SbtOsgi, MultiJvmPlugin)
+  .dependsOn(core % "test->test;compile->compile", session)
+  .settings(
+    libraryDependencies ++= Dependencies.reconcilerDependencies
```
Comment (patriknw) [inline build.sbt]: set `name` and testOptions as in above `core` project? also Automatic-Module-Name

File: core/src/main/scala/akka/persistence/cassandra/Extractors.scala [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+          None
+        }
+
+        Future.successful(
+          RawEvent(
+            sequenceNr,
+            Serialized(
+              row.getString("persistence_id"),
+              row.getLong("sequence_nr"),
+              row.getByteBuffer("event"),
```
Comment (patriknw) [inline core/src/main/scala/akka/persistence/cassandra/Extractors.scala]: in other places we also check for the old "message" column. See usages of `eventDeserializer.columnDefinitionCache.hasMessageColumn`

File: core/src/test/resources/logback-test.xml [hunk 3]
```diff
@@ -1,19 +1,17 @@
...
-    <root level="INFO">
-        <appender-ref ref="CONSOLE"/>
-    </root>
-
+  <appender name="CONSOLE"
+  class="ch.qos.logback.core.ConsoleAppender">
+    <target>System.out</target>
+    <encoder>
+      <pattern>%date{MM/dd HH:mm:ss} %-5level[%thread] %logger{1}
+      [%X{akkaSource}] - %m%n%xException</pattern>
```
Comment (patriknw) [inline core/src/test/resources/logback-test.xml]: does that result in line breaks in the log messages?

File: reconciler/src/main/scala/akka/persistence/cassandra/reconciler/AllTags.scala [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+@InternalApi
+private[akka] final class AllTags(session: ReconciliationSession) {
+
+  def execute(): Source[String, NotUsed] = {
+    session
+      .selectAllTagProgress()
+      .map(_.getString("tag"))
+      .statefulMapConcat(() => {
+        tag =>
+          var seen = Set.empty[String]
```
Comment (patriknw) [inline reconciler/src/main/scala/akka/persistence/cassandra/reconciler/AllTags.scala]: shouldn't this `var` be outside the per element function?

File: reconciler/src/main/scala/akka/persistence/cassandra/reconciler/Reconciliation.scala [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+import java.util.UUID
+import java.lang.{ Long => JLong }
+
+/**
+ * Database actions for reconciliation
+ *
+ * INTERNAL API
+ */
+@InternalApi
+final class ReconciliationSession(session: CassandraSession, statements: CassandraStatements)(
```
Comment (patriknw) [inline reconciler/src/main/scala/akka/persistence/cassandra/reconciler/Reconciliation.scala]: private[akka]

File: reconciler/src/test/scala/doc/reconciler/ReconciliationCompileOnly.scala [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import akka.persistence.cassandra.reconciler.Reconciliation
+import akka.actor.ActorSystem
+import scala.concurrent.Future
+import akka.Done
+//#imports
+
+class ReconciliationCompileOnly {
+
+  //#reconcile
+  // System should have the same configuration as your application
```
Comment (patriknw) [inline reconciler/src/test/scala/doc/reconciler/ReconciliationCompileOnly.scala]: maybe clarify that it shouldn't join cluster and such

**Comment by chbatey** [comment]
Needs to separate out a java/scaladsl or maybe not worth  it?

File: reconciler/src/main/scala/akka/persistence/cassandra/reconciler/AllTags.scala (hunk 4 above)
Comment (chbatey) [reply reconciler/src/main/scala/akka/persistence/cassandra/reconciler/AllTags.scala]: /me looks at his tests

File: core/src/test/resources/logback-test.xml (hunk 3 above)
Comment (chbatey) [reply core/src/test/resources/logback-test.xml]: Not sure but will revert, some auto . formatting some where

**Comment by patriknw** [comment]
> Needs to separate out a java/scaladsl or maybe not worth it?

It's the futures and such in the extension? Should be fairly simple delegation/conversion. We can merge this first and do that separately.

**Comment by chbatey** [comment]
Yep for the futures and streams

File: core/src/main/scala/akka/persistence/cassandra/Extractors.scala (hunk 2 above)
Comment (chbatey) [reply core/src/main/scala/akka/persistence/cassandra/Extractors.scala]: I guess we should support data from that far back,  means we it also needs to know about serialization

**Pull request merged**